node/core/approval-voting/src/criteria.rs

```diff
@@ -483,52 +668,115 @@ pub(crate) fn check_assignment_cert(
+
+	match &assignment.kind {
+		AssignmentCertKindV2::RelayVRFModuloCompact { core_bitfield } => {
+			// Check that claimed core bitfield match the one from certificate.
+			if &claimed_core_indices != core_bitfield {
+				return Err(InvalidAssignment(Reason::VRFModuloCoreIndexMismatch))
+			}
+
+			let (vrf_in_out, _) = public
+				.vrf_verify_extra(
+					relay_vrf_modulo_transcript_v2(relay_vrf_story),
+					&vrf_output.0,
+					&vrf_proof.0,
+					assigned_cores_transcript(core_bitfield),
+				)
+				.map_err(|_| InvalidAssignment(Reason::VRFModuloOutputMismatch))?;
+
+			let resulting_cores = relay_vrf_modulo_cores(
+				&vrf_in_out,
+				config.relay_vrf_modulo_samples,
+				config.n_cores,
+			);
+
+			// TODO: Enforce that all claimable cores are claimed, or include refused cores.
+			// Currently validators can opt out of checking specific cores.
+			// This is the same issue to how validator can opt out and not send their assignments in the first place.
+
+			// Ensure that the `vrf_in_out` actually includes all of the claimed cores.
+			if claimed_core_indices.iter_ones().fold(true, |cores_match, core| {
+				cores_match & resulting_cores.contains(&CoreIndex(core as u32))
+			}) {
+				Ok(0)
+			} else {
+				gum::debug!(
+					target: LOG_TARGET,
+					?resulting_cores,
+					?claimed_core_indices,
+					"Assignment claimed cores mismatch",
+				);
+				Err(InvalidAssignment(Reason::VRFModuloCoreIndexMismatch))
+			}
+		},
+		AssignmentCertKindV2::RelayVRFModulo { sample } => {
+			if *sample >= config.relay_vrf_modulo_samples {
 				return Err(InvalidAssignment(Reason::SampleOutOfBounds))
 			}
 
 			let (vrf_in_out, _) = public
 				.vrf_verify_extra(
-					relay_vrf_modulo_transcript(relay_vrf_story, sample),
-					&vrf_signature.output.0,
-					&vrf_signature.proof.0,
-					assigned_core_transcript(claimed_core_index),
+					relay_vrf_modulo_transcript_v1(relay_vrf_story, *sample),
+					&vrf_output.0,
+					&vrf_proof.0,
+					assigned_core_transcript(CoreIndex(
+						claimed_core_indices.first_one().expect("Checked above; qed") as u32,
+					)),
 				)
 				.map_err(|_| InvalidAssignment(Reason::VRFModuloOutputMismatch))?;
 
+			let core = relay_vrf_modulo_core(&vrf_in_out, config.n_cores);
 			// ensure that the `vrf_in_out` actually gives us the claimed core.
-			if relay_vrf_modulo_core(&vrf_in_out, config.n_cores) == claimed_core_index {
+			if core.0 as usize == claimed_core_indices.first_one().expect("Checked above; qed") {
 				Ok(0)
 			} else {
+				gum::debug!(
+					target: LOG_TARGET,
+					?core,
+					?claimed_core_indices,
+					"Assignment claimed cores mismatch",
+				);
 				Err(InvalidAssignment(Reason::VRFModuloCoreIndexMismatch))
 			}
 		},
-		AssignmentCertKind::RelayVRFDelay { core_index } => {
-			if core_index != claimed_core_index {
+		AssignmentCertKindV2::RelayVRFDelay { core_index } => {
+			if core_index.0 as usize !=
+				claimed_core_indices.first_one().expect("Checked above; qed")
```

Comment:
This doesn't enforce that `count_ones() == 1`, only that the first one is the claimed. So `(1001, CoreIndex(0)` would pass but `(1001, CoreIndex(4))` would fail

Reply:
good point!

Reply:
fixed